x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -71,63 +75,86 @@
 
 public class LoggingAuditTrail extends AbstractComponent implements AuditTrail, ClusterStateListener {
 
+    // WARNING: changing any of this requires changing the log4j2.properties file too
```

Comment:
> What is the behavior if I use the wrong name in log4j2.properties and try to add log message, does it fail for the field not found or it is ignored?

All key references to the `StringMapMessage` dictionary are wrapped in the log4j2 function `%varsNotEmpty{}`. Because of this, missing fields will not be printed. So, if you change the name here but not in the properties file (or vice-versa), the field will be silently ignored.

> I am not sure what the purpose is for StringMapMessage#validate(key, value) (by default it does not have any validation) something that could be used to enforce validation on field, I do not feel strongly about this as this adds overhead to the logging but still raising it so the we do not miss it.

That's an interesting idea, but I don't think `StringMapMessage#validate(key, value)` can help us achieve this validation. That's because `StringMapMessage` building is oblivious of the layout.
What we could try instead, is to get a hold of the layout at runtime, get the pattern string and check if each map references in the pattern layout is backed by a field we add to the `StringMapMessage`.
Might turn out to be too complicated, but I will try and report back.

Reply:
@bizybot I've been mulling over your suggestion to have some sort of validation of the pairings between field names in the java code and field names in the properties file. I have pushed https://github.com/elastic/elasticsearch/pull/31931/commits/2a3857db5f9a54c53817c077df155846838e1b0f specifically in relation to this. What follows is my reasoning:

First off, the properties file is there to allow the administrator to configure the format. For example it is perfectly reasonable to expect that some fields will be disabled by the administrator by changing the format. However, it is true that a typo might slip in and some field will not be printed, as otherwise expected. In this case, maybe an ERROR log entry (in the usual log) would be a nice to have; or even refusing to apply the config change. However, this is not that simple in practice, mainly because there is no such feedback loop from the log4j library itself. Also, it is too complex to parse the properties file ourselves because it might contain what's called a [Pattern Selector](https://logging.apache.org/log4j/2.x/manual/layouts.html#PatternLayout). This is basically a more versatile pattern, complete with scripting to select the pattern and what not...

Therefore I have abandoned the holy grail of trying to protected from administrator errors and aimed to solve the simpler problem of protecting against our own mistakes. In this respect, I have also avoided parsing the pattern from the config and pairing each variable reference to a field in the code (this would have been another test method). The reason for this is that it turned out to be messy and brittle: there are quite some functions available and the pattern can get complex. In addition it required to have a list of all the fields on the java code; just messy...

What I have decided to go with is to have the `LoggingAuditTrailTests` pull the layout that they're using from the actual `log4j2.properties` file. Now, if we ship with an error in the config some tests will catch it. Moreover, this change avoids the pattern duplication that was used previously (the tests programmatically configure a Mockup appender with the legit layout). Furthermore, having tests run over the true formatting, acknowledges the fact that formatting is split between the code and log4j (https://github.com/elastic/elasticsearch/pull/31931#discussion_r215213222) and we're testing the final composed format of the log entries.

Hope this assuages your worries about config file discordance.
Let me know what you think!